rust/arrow/src/array/union.rs

```diff
@@ -0,0 +1,1172 @@
+//! types are also supported but not via `UnionBuilder`, see the tests for examples.
+//!
+//! # Example: Dense Memory Layout
+//!
+//! ```
+//! use arrow::array::UnionBuilder;
+//! use arrow::datatypes::{Float64Type, Int32Type};
+//!
+//! # fn main() -> arrow::error::Result<()> {
+//! let mut builder = UnionBuilder::new_dense(3);
+//! builder.append::<Int32Type>("a", 1).unwrap();
+//! builder.append::<Float64Type>("b", 3.0).unwrap();
+//! builder.append::<Int32Type>("a", 4).unwrap();
+//! let union = builder.build().unwrap();
+//!
+//! assert_eq!(union.type_id(0), 0_i8);
+//! assert_eq!(union.type_id(1), 1_i8);
+//! assert_eq!(union.type_id(2), 0_i8);
+//!
+//! assert_eq!(union.value_offset(0), 0_i32);
+//! assert_eq!(union.value_offset(1), 0_i32);
+//! assert_eq!(union.value_offset(2), 1_i32);
+//!
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! # Example: Sparse Memory Layout
+//! ```
+//! use arrow::array::UnionBuilder;
+//! use arrow::datatypes::{Float64Type, Int32Type};
+//!
+//! # fn main() -> arrow::error::Result<()> {
+//! let mut builder = UnionBuilder::new_sparse(3);
+//! builder.append::<Int32Type>("a", 1).unwrap();
+//! builder.append::<Float64Type>("b", 3.0).unwrap();
+//! builder.append::<Int32Type>("a", 4).unwrap();
+//! let union = builder.build().unwrap();
+//!
+//! assert_eq!(union.type_id(0), 0_i8);
+//! assert_eq!(union.type_id(1), 1_i8);
+//! assert_eq!(union.type_id(2), 0_i8);
+//!
+//! assert_eq!(union.value_offset(0), 0_i32);
+//! assert_eq!(union.value_offset(1), 1_i32);
+//! assert_eq!(union.value_offset(2), 2_i32);
+//!
+//! # Ok(())
+//! # }
+//! ```
+use crate::array::{
+    builder::{builder_to_mutable_buffer, mutable_buffer_to_builder, BufferBuilderTrait},
+    make_array, Array, ArrayData, ArrayDataBuilder, ArrayDataRef, ArrayRef,
+    BooleanBufferBuilder, BufferBuilder, Int32BufferBuilder, Int8BufferBuilder,
+};
+use crate::buffer::{Buffer, MutableBuffer};
+use crate::datatypes::*;
+use crate::error::{ArrowError, Result};
+
+use crate::util::bit_util;
+use core::fmt;
+use std::any::Any;
+use std::collections::HashMap;
+use std::mem::size_of;
+
+/// An Array that can represent slots of varying types
+pub struct UnionArray {
+    data: ArrayDataRef,
+    boxed_fields: Vec<ArrayRef>,
+}
+
+impl UnionArray {
+    /// Creates a new `UnionArray`.
+    ///
+    /// Accepts type ids, child arrays and optionally offsets (for dense unions) to create
+    /// a new `UnionArray`.  This method makes no attempt to validate the data provided by the
+    /// caller and assumes that each of the components are correct and consistent with each other.
+    /// See `try_new` for an alternative that validates the data provided.
+    ///
+    /// # Data Consistency
+    ///
+    /// The `type_ids` `Buffer` should contain `i8` values.  These values should be greater than
+    /// zero and must be less than the number of children provided in `child_arrays`.  These values
```

Comment:
Note that the Union type in C++ and in the IPC format allows you to specify logical type ids.
For example if creating `UnionType({float32, int32}, type_ids=[5, 7])`, then a UnionArray's type ids buffer could be something like `[5, 7, 7]`.

Reply:
I was just going but [this](https://arrow.apache.org/docs/format/Columnar.html#union-layout).

It's not immediately obvious why you would need that.  I'll have to study the C++ implementation.  I'll open a follow up PR.